#! /usr/bin/env python

PACKAGE="uvc_ros_driver"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

#general parameters
gen.add("CAMERA_AUTOEXP", bool_t, 0, "Enable auto exposure AEC function.", True)
gen.add("CAMERA_MIN_E", int_t, 0, "Minimum exposure time when AEC enabled", 0, 0, 480)
gen.add("CAMERA_MAX_E", int_t, 0, "Maximum exposure time when AEC enabled", 378, 0, 480)
gen.add("CAMERA_AUTOG", bool_t, 0, "Enable auto exposure AEC function.", True)
gen.add("PRIMARY_CAM_MODE", bool_t, 0, "Enable primary camera pair mode.", False)
gen.add("CAMERA_TILE", bool_t, 0, "Enable camera tile mode", False)
gen.add("STEREO_BAYER_D", bool_t, 0, "Disable bayer pattern removal", False)
gen.add("ADIS_IMU", bool_t, 0, "Enable the external IMU.", False);


#flip parameters for all 10 cameras
gen.add("CAMERA_0_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_0_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_1_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_1_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_2_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_2_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_3_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_3_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_4_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_4_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_5_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_5_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_6_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_6_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_7_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_7_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_8_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_8_VFLIP", bool_t, 0, "Enable vertical image flip.", False)
gen.add("CAMERA_9_HFLIP", bool_t, 0, "Enable horizontal image flip", False)
gen.add("CAMERA_9_VFLIP", bool_t, 0, "Enable vertical image flip.", False)

#exposure parameters for all 10 cameras
gen.add("CAMERA_EXP_CAM0", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM1", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM2", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM3", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM4", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM5", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM6", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM7", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM8", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)
gen.add("CAMERA_EXP_CAM9", int_t, 0, "Exposure time when AEC disabled.", 40, 0, 480)

gen.add("CAMERA_G_CAM0", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM1", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM2", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM3", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM4", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM5", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM6", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM7", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM8", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)
gen.add("CAMERA_G_CAM9", int_t, 0, "Gain when AEG disabled.", 16, 16, 64)

#stereo parameter for all 5 pairs
gen.add("STEREO_TH_CAM1", int_t, 0, "Disparity matching threshold value", 140, 0, 255)
gen.add("STEREO_LR_CAM1", int_t, 0, "Disparity matching LR check value", 3, 0, 31)
gen.add("STEREO_OF_CAM1", int_t, 0, "Disparity range offset", 0, 0, 127)
gen.add("STEREO_P1_CAM1", int_t, 0, "SGM penalty cost 1", 10, 0, 127)
gen.add("STEREO_P2_CAM1", int_t, 0, "SGM penalty cost2", 250, 0, 255)
gen.add("STEREO_RE_CAM1", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_CE_CAM1", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_FP_CAM1", bool_t, 0, "Switch left and right camera inputs", False)

gen.add("STEREO_TH_CAM3", int_t, 0, "Disparity matching threshold value", 140, 0, 255)
gen.add("STEREO_LR_CAM3", int_t, 0, "Disparity matching LR check value", 3, 0, 31)
gen.add("STEREO_OF_CAM3", int_t, 0, "Disparity range offset", 0, 0, 127)
gen.add("STEREO_P1_CAM3", int_t, 0, "SGM penalty cost 1", 10, 0, 127)
gen.add("STEREO_P2_CAM3", int_t, 0, "SGM penalty cost2", 250, 0, 255)
gen.add("STEREO_RE_CAM3", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_CE_CAM3", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_FP_CAM3", bool_t, 0, "Switch left and right camera inputs", False)

gen.add("STEREO_TH_CAM5", int_t, 0, "Disparity matching threshold value", 140, 0, 255)
gen.add("STEREO_LR_CAM5", int_t, 0, "Disparity matching LR check value", 3, 0, 31)
gen.add("STEREO_OF_CAM5", int_t, 0, "Disparity range offset", 0, 0, 127)
gen.add("STEREO_P1_CAM5", int_t, 0, "SGM penalty cost 1", 10, 0, 127)
gen.add("STEREO_P2_CAM5", int_t, 0, "SGM penalty cost2", 250, 0, 255)
gen.add("STEREO_RE_CAM5", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_CE_CAM5", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_FP_CAM5", bool_t, 0, "Switch left and right camera inputs", False)

gen.add("STEREO_TH_CAM7", int_t, 0, "Disparity matching threshold value", 140, 0, 255)
gen.add("STEREO_LR_CAM7", int_t, 0, "Disparity matching LR check value", 3, 0, 31)
gen.add("STEREO_OF_CAM7", int_t, 0, "Disparity range offset", 0, 0, 127)
gen.add("STEREO_P1_CAM7", int_t, 0, "SGM penalty cost 1", 10, 0, 127)
gen.add("STEREO_P2_CAM7", int_t, 0, "SGM penalty cost2", 250, 0, 255)
gen.add("STEREO_RE_CAM7", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_CE_CAM7", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_FP_CAM7", bool_t, 0, "Switch left and right camera inputs", False)

gen.add("STEREO_TH_CAM9", int_t, 0, "Disparity matching threshold value", 140, 0, 255)
gen.add("STEREO_LR_CAM9", int_t, 0, "Disparity matching LR check value", 3, 0, 31)
gen.add("STEREO_OF_CAM9", int_t, 0, "Disparity range offset", 0, 0, 127)
gen.add("STEREO_P1_CAM9", int_t, 0, "SGM penalty cost 1", 10, 0, 127)
gen.add("STEREO_P2_CAM9", int_t, 0, "SGM penalty cost2", 250, 0, 255)
gen.add("STEREO_RE_CAM9", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_CE_CAM9", bool_t, 0, "Enable disparity output.", False)
gen.add("STEREO_FP_CAM9", bool_t, 0, "Switch left and right camera inputs", False)

exit(gen.generate(PACKAGE, "uvc_ros_driver", "UvcDriver"))
